backend/wmg/pipeline/summary_cubes/calculate_markers.py

```diff
@@ -145,12 +115,120 @@ def _query_tiledb(
     agg = query.groupby(gb_dims_es).sum()
     n_cells = cell_counts_query.groupby(gb_dims).sum()["n_total_cells"]
 
-    keep_dataset_ids = group_by_dims is None and "dataset_id" in filters.keys()
-    if genes is None:
-        genes = list(agg.index.levels[0])
+    genes = list(agg.index.levels[0])
+    n_cells_per_gene, n_cells_index = _calculate_true_n_cells(n_cells, genes, dataset_to_gene_ids, keep_dataset_ids)
+
+    # roll up across cell type descendants
+    if "cell_type_ontology_term_id" in agg.index.names:
+        # make tidy
+        for name in agg.index.names:
+            agg[name] = agg.index.get_level_values(name)
+        agg = rollup_across_cell_type_descendants(agg)
+        # remove tidy
+        agg.drop(columns=agg.index.names, inplace=True)
+
+    return agg, n_cells_per_gene, n_cells_index, genes
+
+
+def _query_tiledb_context(
+    target_filters: dict,
+    context_filters: dict,
+    corpus: Optional[Union[WmgSnapshot, str]] = None,
+):
+    """
+    Query the tileDB array and return required data artifacts for downstream processing.
+
+    Arguments
+    ---------
+    target_filters - dict,
+        Dictionary of filters for the target population
 
-    t_n_cells_sum = _calculate_true_n_cells(n_cells, genes, dataset_to_gene_ids, keep_dataset_ids)
-    return agg, t_n_cells_sum, genes
+    context_filters - dict,
+        Dictionary of filters for the context population
+
+    corpus - str or WmgSnapshot, optional, default None
+        If string, it is the path to the snapshot.
+        If WmgSnapshot, it is the snapshot object.
+        If None, the snapshot will be fetched from AWS.
+
+    Returns
+    -------
+    - agg: query result aggregated by filter keys
+    - n_cells_context: Series of aggregated cell counts from the context query further
+        stratified by dataset IDs.
+    - n_cells_per_gene: numpy array
+        Each row in the array contains the true number of cells present for the combination of
+        filters in the context query.
+    - n_cells_index: pandas MultiIndex
+        Each element in the index contains the combination of filters for the corresponding row in
+        `n_cells_per_gene`
+    - genes: a list of all genes with nonzero expression found in the query result
+    """
+    group_by_dims = list(target_filters.keys())
+    keep_dataset_ids = "dataset_ids" in target_filters
+    return _query_tiledb_context_memoized(context_filters, group_by_dims, keep_dataset_ids, corpus=corpus)
+
+
+def _query_tiledb_target(
+    target_filters: dict,
+    context_filters: dict,
+    context_agg: pd.DataFrame,
+    n_cells_per_gene_context: np.array,
+    n_cells_index_context: pd.MultiIndex,
+):
```

Comment:
This function extracts the data for the target population from the context query results.